// @skpm/builder/lib/utils/webpackConfig.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

let getCommands = (() => {
  var _ref = _asyncToGenerator(function* (output, entry, options) {
    return Promise.all(entry.identifiers.map(function (commandIdentifier) {
      return (0, _webpackCommandPlugin2.default)(output, commandIdentifier, options);
    }));
  });

  return function getCommands(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
})();

// avoid looking it up every time


exports.default = getWebpackConfig;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpackMerge = require('webpack-merge');

var _webpackMerge2 = _interopRequireDefault(_webpackMerge);

var _terserWebpackPlugin = require('terser-webpack-plugin');

var _terserWebpackPlugin2 = _interopRequireDefault(_terserWebpackPlugin);

var _webpackCommandPlugin = require('./webpackCommandPlugin');

var _webpackCommandPlugin2 = _interopRequireDefault(_webpackCommandPlugin);

var _webpackHeaderFooterPlugin = require('./webpackHeaderFooterPlugin');

var _webpackHeaderFooterPlugin2 = _interopRequireDefault(_webpackHeaderFooterPlugin);

var _babelLoader = require('./babelLoader');

var _babelLoader2 = _interopRequireDefault(_babelLoader);

var _resourceLoader = require('./resourceLoader');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const CORE_MODULES = ['buffer', 'console', 'events', 'os', 'path', 'process', 'querystring', 'stream', 'string_decoder', 'timers', 'util'];

const { NODE_ENV } = process.env;
const isProd = NODE_ENV === 'production';

function getWebpackConfig(argv, output, manifestFolder, skpmConfig) {
  let userDefinedWebpackConfig;
  try {
    if (_fs2.default.existsSync(_path2.default.join(process.cwd(), 'webpack.skpm.config.js'))) {
      userDefinedWebpackConfig = require(_path2.default.join(process.cwd(), 'webpack.skpm.config.js'));
    }
  } catch (err) {
    console.error(`${_chalk2.default.red('error')} Error while reading webpack.skpm.config.js`);
    console.error(err);
    process.exit(1);
  }

  const babelLoader = (0, _babelLoader2.default)(skpmConfig);

  return (() => {
    var _ref2 = _asyncToGenerator(function* (entry) {
      let plugins = [new _webpack2.default.EnvironmentPlugin({
        NODE_ENV: NODE_ENV || 'development' // default to 'development'
      })];
      const rules = [babelLoader];

      if (entry.isPluginCommand) {
        if (entry.handlers.find(function (k) {
          return k === '__skpm_run';
        })) {
          console.error(`${_chalk2.default.red('error')} __skpm_run is a reserved keyword for handlers. Please use something else.`);
          process.exit(1);
        }

        plugins.push(new _webpack2.default.ProvidePlugin({
          fetch: require.resolve('sketch-polyfill-fetch'),
          FormData: require.resolve('sketch-polyfill-fetch/lib/form-data'),
          Promise: require.resolve('@skpm/promise')
        }), new _webpackHeaderFooterPlugin2.default(entry.handlers));

        rules.push(_resourceLoader.commandResourceLoader);
        rules.push(_resourceLoader.nibLoader);
        rules.push(_resourceLoader.xcodeprojLoader);
      }

      if (argv.run && entry.isPluginCommand) {
        plugins = plugins.concat((yield getCommands(output, entry, _extends({}, argv, {
          pluginIdentifier: skpmConfig && skpmConfig.identifier
        }))));
      }

      let webpackConfig = _extends({
        mode: isProd ? 'production' : 'development',
        devtool: isProd ? 'none' : 'source-map',
        optimization: {
          minimizer: [new _terserWebpackPlugin2.default({
            terserOptions: {
              mangle: {
                // @see https://bugs.webkit.org/show_bug.cgi?id=171041
                // @see https://github.com/mishoo/UglifyJS2/issues/1753#issuecomment-324814782
                safari10: true
              },
              output: {
                safari10: true
              }
            }
          })]
        },
        module: {
          rules
        },
        resolve: {
          mainFields: ['sketch', 'browser', 'module', 'main'],
          aliasFields: ['sketch', 'browser'],
          extensions: ['.sketch.js', '.js'],
          modules: ['node_modules', _path2.default.join(__dirname, '..', '..', 'node_modules')]
        },
        resolveLoader: {
          modules: [_path2.default.join(__dirname, '..', '..', 'node_modules'), 'node_modules']
        },
        entry: entry.absolutePath,
        externals: [function (context, request, callback) {
          // we only want to mess with pluginCommands
          if (!entry.isPluginCommand) {
            return callback();
          }

          // sketch API
          if (/^sketch\//.test(request) || request === 'sketch') {
            return callback(null, `commonjs ${request}`);
          }
          // core modules shipped in Sketch
          if (CORE_MODULES.indexOf(request) !== -1) {
            return callback(null, `commonjs ${request}`);
          }
          return callback();
        }],
        output: {
          filename: entry.script.replace(/\.(?!js)|\//g, '_'),
          library: entry.isPluginCommand ? 'exports' : undefined,
          path: entry.isPluginCommand ? _path2.default.join(output, 'Contents', 'Sketch') : _path2.default.join(output, 'Contents', 'Resources')
        },
        plugins
      }, entry.isPluginCommand ? { node: false } : null);

      if (userDefinedWebpackConfig) {
        const resolvedUserDefinedConfig = yield userDefinedWebpackConfig(webpackConfig, entry);
        if (resolvedUserDefinedConfig) {
          webpackConfig = _webpackMerge2.default.smart(webpackConfig, resolvedUserDefinedConfig);
        }
      }

      return webpackConfig;
    });

    function webpackConfigGenerator(_x4) {
      return _ref2.apply(this, arguments);
    }

    return webpackConfigGenerator;
  })();
}
